drivers/dstate.c

```diff
@@ -1002,3 +1002,163 @@ void device_alarm_commit(const int device_number)
+			return -1;
+		}
+
+		int bufrw_max = sizeof(buf) - xput_prefix_len;
+		if (bufrw_max <= 15) {
+			/* We need to append max ~13 chars per below, so far */
+			upsdebugx(0, "set_phasecount(): Bad xput_prefix was passed: it is too long - function skipped");
+			return -1;
+		}
+		strncpy(buf, xput_prefix, sizeof(buf));
+		char *bufrw_ptr = buf + xput_prefix_len ;
+
+		/* We either have defined and non-zero (numeric) values below, or NULLs */
+		strncpy(bufrw_ptr, "L1.voltage", bufrw_max);
+		if ((v1 = dstate_getinfo(buf)))    { if (v1[0]  == '0') { v1  = NULL; } }
+		strncpy(bufrw_ptr, "L2.voltage", bufrw_max);
+		if ((v2 = dstate_getinfo(buf)))    { if (v2[0]  == '0') { v2  = NULL; } }
+		strncpy(bufrw_ptr, "L3.voltage", bufrw_max);
+		if ((v3 = dstate_getinfo(buf)))    { if (v3[0]  == '0') { v3  = NULL; } }
+		strncpy(bufrw_ptr, "L1-N.voltage", bufrw_max);
+		if ((v1n = dstate_getinfo(buf)))   { if (v1n[0] == '0') { v1n = NULL; } }
+		strncpy(bufrw_ptr, "L2-N.voltage", bufrw_max);
+		if ((v2n = dstate_getinfo(buf)))   { if (v2n[0] == '0') { v2n = NULL; } }
+		strncpy(bufrw_ptr, "L3-N.voltage", bufrw_max);
+		if ((v3n = dstate_getinfo(buf)))   { if (v3n[0] == '0') { v3n = NULL; } }
+		strncpy(bufrw_ptr, "L1-L2.voltage", bufrw_max);
+		if ((v12 = dstate_getinfo(buf)))   { if (v12[0] == '0') { v12 = NULL; } }
+		strncpy(bufrw_ptr, "L2-L3.voltage", bufrw_max);
+		if ((v23 = dstate_getinfo(buf)))   { if (v23[0] == '0') { v23 = NULL; } }
+		strncpy(bufrw_ptr, "L3-L1.voltage", bufrw_max);
+		if ((v31 = dstate_getinfo(buf)))   { if (v31[0] == '0') { v31 = NULL; } }
+		strncpy(bufrw_ptr, "L3.current", bufrw_max);
+		if ((c1 = dstate_getinfo(buf)))    { if (c1[0]  == '0') { c1  = NULL; } }
+		strncpy(bufrw_ptr, "L2.current", bufrw_max);
+		if ((c2 = dstate_getinfo(buf)))    { if (c2[0]  == '0') { c2  = NULL; } }
+		strncpy(bufrw_ptr, "L3.current", bufrw_max);
+		if ((c3 = dstate_getinfo(buf)))    { if (c3[0]  == '0') { c3  = NULL; } }
+		strncpy(bufrw_ptr, "voltage", bufrw_max);
+		if ((v0 = dstate_getinfo(buf)))    { if (v0[0]  == '0') { v0  = NULL; } }
+		strncpy(bufrw_ptr, "current", bufrw_max);
+		if ((c0 = dstate_getinfo(buf)))    { if (c0[0]  == '0') { c0  = NULL; } }
+
+		if ( (v1 && v2 && v3) ||
+		     (v1n && v2n && v3n) ||
+		     (c1 && (c2 || c3)) ||
+		     (c2 && (c1 || c3)) ||
+		     (c3 && (c1 || c2)) ||
+		     v12 || v23 || v31 ) {
+			upsdebugx(5, "set_phasecount(): determined a 3-phase case");
+			*num_phases = 3;
+			*inited_phaseinfo = 1;
+			detected_phaseinfo = 1;
+		} else if ( /* We definitely have only one non-zero line */
+		     !v12 && !v23 && !v31 && (
+		     (c0 && !c1 && !c2 && !c3) ||
+		     (v0 && !v1 && !v2 && !v3) ||
+		     (c1 && !c2 && !c3) ||
+		     (!c1 && c2 && !c3) ||
+		     (!c1 && !c2 && c3) ||
+		     (v1 && !v2 && !v3) ||
+		     (!v1 && v2 && !v3) ||
+		     (!v1 && !v2 && v3) ||
+		     (v1n && !v2n && !v3n) ||
+		     (!v1n && v2n && !v3n) ||
+		     (!v1n && !v2n && v3n) ) ) {
```

Comment:
Interesting point... initially (in earlier commits) the logic was all around voltages. From what I gather, on a 3-phase capable device with just 1 phase connected to active load (others not connected or loads not powered on), there is still a voltage reading but negligible/zero current (and realpower, probably).

So my checks revolved around devices reporting non-zero value on just one of the line-metrics (since a single-phase load might be wired to any one phase).

The tests above for `!c0 && !v0` are not correct however, because in the XML for a 3-phase capable device I see both an overall metric and per-phase values - so for the disbalanced usecase (or one reported by a general form of 3-phase markup) we'd see both `v0` and one of those other voltages.

Otherwise, your form (and comment) gives good food for thought, maybe I'll get back to think it over after some rest (e.g. adopt this solution, or really prove it wrong), or maybe someone else will :)